core/src/main/java/com/datastax/oss/driver/api/core/paging/OffsetPager.java

```diff
@@ -0,0 +1,316 @@
+ * small result set.
+ *
+ * <p>On the other hand, if you want to display 10,000 elements per page and allow the user to query
+ * page 7,384, this class is not going to work. You should actively defend against such aberrant
+ * values, see the next paragraph.
+ *
+ * <h3>Establishing application-level guardrails</h3>
+ *
+ * If you use this class, we strongly recommend that you implement hard limits in your application
+ * code. If the page number is exposed to the user (for example if it is passed as a URL parameter),
+ * make sure it is properly validated and enforce a maximum, so that an attacker can't inject a
+ * large value that could potentially fetch millions of rows.
+ *
+ * <h3>Relation with protocol-level paging</h3>
+ *
+ * There is also something called "paging" at the protocol level (controlled by the config option
+ * {@code basic.request.page-size}). It simply means that the driver might receive the results in
+ * multiple "chunks"; this happens under the hood and is completely transparent for users of this
+ * class. You don't need to set the protocol page size to the same value as the logical page size.
+ */
+@ThreadSafe
+public class OffsetPager {
+
+  /** A page returned as the result of an offset query. */
+  public interface Page<ElementT> {
+
+    /** The elements in the page. */
+    @NonNull
+    List<ElementT> getElements();
+
+    /**
+     * The page number (1 for the first page, 2 for the second page, etc).
+     *
+     * <p>Note that it may be different than the requested page number: if the result set is too
+     * short, {@link OffsetPager#getPage(PagingIterable, int)} returns the last page instead, and
+     * this method will return the number of that last page.
+     */
+    int getPageNumber();
+
+    /** Whether this is the last page in the result set. */
+    boolean isLast();
+  }
+
+  private final int pageSize;
+
+  /**
+   * Creates a new instance.
+   *
+   * @param pageSize the number of elements per page. Must be greater than or equal to 1.
+   */
+  public OffsetPager(int pageSize) {
+    if (pageSize < 1) {
+      throw new IllegalArgumentException("Invalid pageSize, expected >=1, got " + pageSize);
+    }
+    this.pageSize = pageSize;
+  }
+
+  /**
+   * Skips the beginning of a set of results to return the given page, assuming the given page size.
+   *
+   * @param iterable the iterable to extract the results from: typically a {@link ResultSet}, or a
+   *     {@link PagingIterable} returned by the mapper.
+   * @param targetPageNumber the page to return (1 for the first page, 2 for the second page, etc).
+   *     Must be greater than or equal to 1.
+   * @return the requested page, or the last page if the requested page was past the end of the
+   *     iterable.
+   * @throws IllegalArgumentException if the conditions on the arguments are not respected.
+   */
+  @NonNull
+  public <ElementT> Page<ElementT> getPage(
```

Comment:
can't we reuse the same method for sync and async version?

Reply:
How would that work? `PagingIterable` and `AsyncPagingIterable` are unrelated types.

Reply:
maybe we could extract some common bits into one class but you are right, maybe it is better to keep them totally separate